signup/signup.js

```diff
@@ -13,6 +17,191 @@ $(document).ready(function() {
 
 });
 
+let datesAllowed = [];
+var sitesWithShifts = [];
+let minTime = '0';
+let maxTime = '23';
+
+class SiteWithShift {
+	constructor(siteId, startTime, endTime) {
+		this.siteId = siteId;
+		const date = this.getDate(startTime);
+		this.dates = [];
+		this.dates.push(date);
+		this.shiftTimes = [];
+		this.shiftTimes.push({
+			date: date,
+			startTime: startTime,
+			endTime: endTime
+		});
+	}
+
+	getDate(dateTime) {
+		let t = dateTime.split(/[- :]/);
+		return new Date(t[0],t[1]-1,t[2]);
+	}
+
+	addShift(startTime, endTime) {
+		const date = this.getDate(startTime);
+		this.dates.push(date);
+		this.shiftTimes.push({
+			date: date,
+			startTime: startTime,
+			endTime: endTime
+		});
+	}
+}
+
+function loadAllSites() {
+	var request = $.ajax({
+		url: "/server/api/sites/getAll.php",
+		type: "GET",
+		dataType: "JSON",
+		data: ({
+			"siteId": true,
+			"title": true
+		}),
+		cache: false
+	})
+	.done(function(data) {
+		$siteSelect = $("#sitePicker select");
+		$siteSelect.append($('<option disabled selected value style="display:none"> -- select an option -- </option>'));
+		for(const site of data) {
+			$siteSelect.append($('<option>', {
+				value: site.siteId,
+				text : site.title
+			}));
+		}
+	});
+}
+
+// Load all of the shifts and store in a global variable
+// TODO year
+function loadAllShifts(year=(new Date()).getFullYear()) {
+	var request = $.ajax({
+		url: "/server/api/shifts/getAllShifts.php",
+		type: "GET",
+		dataType: "JSON",
+		data: ({
+			"year": year,
+			"startTime": true,
+			"endTime": true,
+			"siteId": true
+		}),
+		cache: false
+	})
+	.done(function(data) {
+		for(const shift of data) {
+			let sitesArray = sitesWithShifts.map(function(s) {return s.siteId});
+			if(sitesArray.includes(shift.siteId)) {
+				let i = sitesArray.indexOf(shift.siteId);
```

Comment:
I'm a little concerned about what is happening here. This method seems to assume that sitesWithShifts and sitesArray have matching indices for every siteId. I don't think this is an assumption we should make. You may be better off using a Map.